// controllers/yelp.controller.js
// 3rd party modules
var _und = require('underscore');
var _async = require('async');

// GrubGroup defined modules
var _um_utils = require(MODULE_PATH + 'utils');

var yelp_controller = function(response) {
	this.response = response;
	return this;
};

yelp_controller.prototype.response = null;

yelp_controller.prototype.get_sugguestion = function(input) {
	// write headers
	this.response.writeHead(200, {"Content-Type": "application/json"});

	var yelp = require("yelp").createClient({
		consumer_key: "PZSwO6nRWsMXCqX-YdUEJA",
		consumer_secret: "4ptFi4GlR-A-nV-DJ323f7mEDKA",
		token: "nZWejbCEiOCypu7C8f8XvsbdLxv-Uh_G",
		token_secret: "f4gAU_UTewoTJvHllTDJ8Hi6g-k"
	});

	// get list of Yelp categories of restaurants
	var categories = _und.toArray(_um_utils.get_categories());

	// See http://www.yelp.com/developers/documentation/v2/search_api
	var search_restaurant = function (category, callback) {
		yelp.search({term: "", ll: input.latitude + ',' + input.longitude, radius_filter: input.radius, category_filter: category, limit : "20"}, function(error, data) {
			//console.log(error);
			console.log('get ' + category);
			callback(null, data.businesses);
		});
	}

	var that = this;

	_async.map(categories, search_restaurant, function (error, results) {
		results = _und.compact(_und.flatten(results, 'shallow'));	// 1. move all nested objects one level higher, so merge results into one array; remove any nulls or blanks
		results = _und.filter(results, function (place) { return place.rating >= 3; });	// 2. remove any results with less than 3 stars
		results = _und.shuffle(results);	// 3. randomize the order of the objects in the array

		var output = JSON.stringify(results[_und.random(results.length - 1)]);	// 4. get random index of item and display that item

		// check if callback was provided
		if(input.callback == null) that.response.write(output);
		else that.response.write(input.callback + '(' + output + ');');

		that.response.end();
	});
}

module.exports.createClient = function(response) {
	return new yelp_controller(response);
};
